go/test/endtoend/vtgate/foreignkey/stress/fk_stress_test.go

```diff
@@ -0,0 +1,1180 @@
+		if err := clusterInstance.StartKeyspace(*keyspace, []string{"1"}, 1, false); err != nil {
+			return 1, err
+		}
+
+		vtgateInstance := clusterInstance.NewVtgateInstance()
+		// Start vtgate
+		if err := vtgateInstance.Setup(); err != nil {
+			return 1, err
+		}
+		// ensure it is torn down during cluster TearDown
+		clusterInstance.VtgateProcess = *vtgateInstance
+		vtParams = mysql.ConnParams{
+			Host: clusterInstance.Hostname,
+			Port: clusterInstance.VtgateMySQLPort,
+		}
+
+		return m.Run(), nil
+	}()
+	if err != nil {
+		fmt.Printf("%v\n", err)
+		os.Exit(1)
+	} else {
+		os.Exit(exitcode)
+	}
+
+}
+
+func queryTablet(t *testing.T, tablet *cluster.Vttablet, query string, expectError string) *sqltypes.Result {
+	rs, err := tablet.VttabletProcess.QueryTablet(query, keyspaceName, true)
+	if expectError == "" {
+		assert.NoError(t, err)
+	} else {
+		assert.ErrorContains(t, err, expectError)
+	}
+	return rs
+}
+
+func tabletTestName(t *testing.T, tablet *cluster.Vttablet) string {
+	switch tablet {
+	case primary:
+		return "primary"
+	case replica:
+		return "replica"
+	default:
+		assert.FailNowf(t, "unknown tablet", "%v, type=%v", tablet.Alias, tablet.Type)
+	}
+	return ""
+}
+
+func validateReplicationIsHealthy(t *testing.T, tablet *cluster.Vttablet) bool {
+	query := "show replica status"
+	rs, err := tablet.VttabletProcess.QueryTablet(query, keyspaceName, true)
+	assert.NoError(t, err)
+	row := rs.Named().Row()
+	require.NotNil(t, row)
+
+	ioRunning := row.AsString("Replica_IO_Running", "")
+	require.NotEmpty(t, ioRunning)
+	ioHealthy := assert.Equalf(t, "Yes", ioRunning, "Replication is broken. Replication status: %v", row)
+	sqlRunning := row.AsString("Replica_SQL_Running", "")
+	require.NotEmpty(t, sqlRunning)
+	sqlHealthy := assert.Equalf(t, "Yes", sqlRunning, "Replication is broken. Replication status: %v", row)
+
+	return ioHealthy && sqlHealthy
+}
+
+func getTabletPosition(t *testing.T, tablet *cluster.Vttablet) replication.Position {
+	rs := queryTablet(t, tablet, "select @@gtid_executed as gtid_executed", "")
+	row := rs.Named().Row()
+	require.NotNil(t, row)
+	gtidExecuted := row.AsString("gtid_executed", "")
+	require.NotEmpty(t, gtidExecuted)
+	pos, err := replication.DecodePositionDefaultFlavor(gtidExecuted, replication.Mysql56FlavorID)
+	assert.NoError(t, err)
+	return pos
+}
+
+func waitForReplicaCatchup(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	primaryPos := getTabletPosition(t, primary)
+	for {
+		replicaPos := getTabletPosition(t, replica)
+		if replicaPos.GTIDSet.Contains(primaryPos.GTIDSet) {
+			// success
+			return
+		}
+		if !validateReplicationIsHealthy(t, replica) {
+			assert.FailNow(t, "replication is broken; not waiting for catchup")
+			return
+		}
+		select {
+		case <-ctx.Done():
+			assert.FailNow(t, "timeout waiting for replica to catch up")
+			return
+		case <-time.After(time.Second):
+			//
+		}
+	}
+}
```

Comment:
There is a function in `cluster` package which does something very similar. The function can be called like so - `cluster.WaitForReplicationPos(t, primaryTab, replicaTab, "localhost", 60.0)`
That function gets the position for the primary and waits for upto the amount of seconds provided for the replica tablet to be atleast as advanced as that position.

Reply:
OK, converged with `cluster.WaitForReplicationPos`. But to do that, I also made changes to `cluster.WaitForReplicationPos`:

- The function `waitStep()` was incorrect as well as _used_ incorrectly. Removed it and used idiomatic `context.WithCancel` in `WaitForReplicationPos`
- `hostname` is assumed to be `"localhost"` as this is hard coded in the cluster creation anyway.
- Added `validateReplication` function argument. When replication is broken, there is generally no point in waiting for the position. This is opt-in (by default, check is not made).

Reply:
Sweet! Thank-you!